drivers/network/netio/netio.c

```diff
@@ -0,0 +1,903 @@
+/*
+ * PROJECT:     NetIO driver
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     A more convenient networking (TCP/IP, UDP/IP) kernel API (NT6+)
+ * COPYRIGHT:   Copyright 2023-2024 Johannes Thoma <johannes@johannesthoma.com>
+ */
+
+/* We are a NT6+ driver */
+#undef _WIN32_WINNT
+#define _WIN32_WINNT 0x600
+#undef WINVER
+#define WINVER 0x600
+
+/* TODOs:
+    Done: Clean up socket in WskCloseSocket (have RefCount and
+    SocketGet / SocketPut functions).
+    Done: Remove unnecessary code (like Hook of completion)
+    Make it compile with MS VC as well
+    Rebase onto latest master
+    Squash commits (git rebase -i) one for tdihelpers one for netio
+
+    The whole Listen / Accept mechanism is still missing
+    Some minor functions (not used by WinDRBD) are missing
+    (like WskControlClient, WskSocketConnect, ...)
+    Raw sockets are not supported for now.
+    We should have regression tests for that...
+*/
+
+#include <ntdef.h>
+#include <wdm.h>
+#include <wsk.h>
+#include <ndis.h>
+#include <ntifs.h>
+#include <windef.h>
+/* From afd: this should be removed: */
+#include <debug.h>
+
+#include <tdi.h>
+#include <tcpioctl.h>
+#include <tdikrnl.h>
+#include <tdiinfo.h>
+#include <tdi_proto.h>
+#include <tdiconn.h>
+
+/* AFD also defines this. It is used by the tdi helpers */
+DWORD DebugTraceLevel = MIN_TRACE;
+
+#ifndef INVALID_HANDLE_VALUE
+#define INVALID_HANDLE_VALUE ((HANDLE) -1)
```

Comment:
Ok I am not using INVALID_HANDLE_VALUE any more and replaced all occurences with NULL. Was it that what you meant? (INVALID_HANDLE_VALUE is defined as -1 in various ReactOS headers)

Reply:
Yeah that's what I meant, thanks!
`INVALID_HANDLE_VALUE` is an error-prone concept in user mode, because some functions return it and other functions return NULL, and if you check for the wrong one, you might use invalid handles etc.
So introducing that concept into kernel-mode code (where it doesn't natively exist) isn't desirable. We have a few tickets & commits around this, e.g. https://jira.reactos.org/browse/CORE-10187

Reply:
Ok, not using INVALID_HANDLE.